index.js

```diff
@@ -2572,6 +2572,8 @@
       return function(constraints) {
         return function(expTypes) {
           return function(impl) {
+            Object.defineProperty (impl, 'name', {value: name});
+
```

Comment:
Should we be modifying `impl` or the wrapped function? Mutating an argument seems problematic (`impl` could be `Math.sqrt`, for example).